### ISSUE TITLE
Cache cluster topologies

### ISSUE THREAD

**Issue by msau42** [kubernetes-csi/external-provisioner#374]
Once we move CSINode and Node to informers, we could potentially cache the topologies available in the cluster instead of having to List() and iterate through them every time we provision.

/kind feature

**Comment by mucahitkurt** [comment]
@msau42 I think this refers to [these lines of codes](https://github.com/kubernetes-csi/external-provisioner/blob/0c56ab07a68cbe7b186d96f2afa8e00b5895066b/pkg/controller/topology.go#L372-L391)?

If I understand correctly, we would like to cache topology key/value pairs for a cluster, and they are all node labels, so we should cache all node labels and then select according to requested topology keys.

This will make unnecessary to extract requested topology keys from nodes that have at least one requested topology key, we may able to select values of requested topology keys from the informer with one list method call.

But I'm not sure about  benefits vs complexity tradeoff since we already use the informer for Nodes.

**Comment by msau42** [comment]
Yes, basically anytime we have to list, like [here](https://github.com/kubernetes-csi/external-provisioner/blob/0c56ab07a68cbe7b186d96f2afa8e00b5895066b/pkg/controller/topology.go#L321) also.

The performance benefit should theoretically show up when: 1) you have a lot of nodes and 2) you have very few topologies compared to nodes (like zones). But agree, it may be good to write a performance benchmark test first so we can compare the performance when you have to List() nodes every time you provision vs listing topologies from a cache
